File: frontend/types/ensemble.ts
export type EnsembleTradeAction = 'BUY' | 'HOLD' | 'SELL';
export type EnsembleSignalModel = 'fundamental' | 'sentimental' | 'technical';

export interface EnsembleBacktestRequest {
  ticker: string;
  market?: 'US' | 'IN';
  start_date?: string | null;
  end_date?: string | null;
  initial_capital?: number;
  transaction_cost_pct?: number;
  buy_threshold?: number;
  sell_threshold?: number;
  target_long_exposure?: number;
  base_long_exposure?: number;
  technical_exposure_weight?: number;
  fundamental_exposure_weight?: number;
  sentiment_max_exposure?: number;
  min_trade_value?: number;
  min_model_count?: number;
  require_sentiment_signal?: boolean;
  allow_technical_proxy?: boolean;
}

export interface EnsembleModelSignal {
  date: string;
  ticker: string;
  model: EnsembleSignalModel;
  raw_signal: string;
  normalized_score: number;
  confidence: number;
  signal_label?: string | null;
  source: string;
}

export interface EnsembleDecision {
  date: string;
  close: number;
  action: EnsembleTradeAction;
  average_score: number;
  target_exposure?: number | null;
  model_count: number;
  model_scores: Record<string, number>;
  sentiment_action?: EnsembleTradeAction | null;
  support_score: number;
  technical_adjustment: number;
  fundamental_adjustment: number;
}

export interface EnsembleTrade {
  date: string;
  action: EnsembleTradeAction;
  price: number;
  exposure_before: number;
  exposure_after: number;
  trade_value: number;
  transaction_cost: number;
  shares_after: number;
  cash_after: number;
  portfolio_value: number;
}

export interface EnsembleEquityPoint {
  date: string;
  close: number;
  shares: number;
  cash: number;
  exposure: number;
  portfolio_value: number;
  daily_return: number;
}

export interface EnsembleBacktestMetrics {
  initial_capital: number;
  final_value: number;
  total_return_pct: number;
  sharpe_ratio: number;
  max_drawdown_pct: number;
  win_rate_pct: number;
  trade_count: number;
  decision_count: number;
  average_model_count: number;
  coverage_by_model: Record<string, number>;
}

export interface EnsembleBacktestResponse {
  ticker: string;
  market: string;
  start_date: string;
  end_date: string;
  metrics: EnsembleBacktestMetrics;
  decisions: EnsembleDecision[];
  trades: EnsembleTrade[];
  equity_curve: EnsembleEquityPoint[];
  model_signals: EnsembleModelSignal[];
  source_files: string[];
  warnings: string[];
  generated_at: string;
}

File: backend/app/schemas/ensemble.py
from __future__ import annotations

from datetime import date, datetime
from typing import Dict, List, Literal, Optional

from pydantic import BaseModel, Field, model_validator


TradeAction = Literal["BUY", "HOLD", "SELL"]
SignalModel = Literal["fundamental", "sentimental", "technical"]


class EnsembleBacktestRequest(BaseModel):
    ticker: str = Field(..., min_length=1, max_length=20, description="Stock ticker symbol")
    market: Literal["US", "IN"] = Field("US", description="Market: US or India")
    start_date: Optional[date] = Field(None, description="Inclusive backtest start date")
    end_date: Optional[date] = Field(None, description="Inclusive backtest end date")
    initial_capital: float = Field(10000.0, gt=0.0)
    transaction_cost_pct: float = Field(0.001, ge=0.0, le=0.05)
    buy_threshold: float = Field(0.15, ge=-1.0, le=1.0)
    sell_threshold: float = Field(-0.15, ge=-1.0, le=1.0)
    target_long_exposure: float = Field(1.0, ge=0.0, le=2.5)
    base_long_exposure: float = Field(0.6, ge=0.0, le=2.5)
    technical_exposure_weight: float = Field(0.25, ge=0.0, le=2.5)
    fundamental_exposure_weight: float = Field(0.15, ge=0.0, le=2.5)
    sentiment_max_exposure: float = Field(2.5, gt=0.0, le=10.0)
    min_trade_value: float = Field(1.0, ge=0.0)
    min_model_count: int = Field(1, ge=1, le=3)
    require_sentiment_signal: bool = Field(
        True,
        description="Require a sentimental signal before creating a sentiment-led aggregate decision.",
    )
    allow_technical_proxy: bool = Field(
        True,
        description="Use deterministic price-momentum technical proxy if no technical backtest artifact is available.",
    )

    @model_validator(mode="after")
    def validate_thresholds(self) -> "EnsembleBacktestRequest":
        if self.sell_threshold >= self.buy_threshold:
            raise ValueError("sell_threshold must be lower than buy_threshold")
        if self.start_date and self.end_date and self.start_date > self.end_date:
            raise ValueError("start_date must be on or before end_date")
        if self.base_long_exposure > self.target_long_exposure:
            raise ValueError("base_long_exposure must be less than or equal to target_long_exposure")
        return self


class EnsembleModelSignal(BaseModel):
    date: date
    ticker: str
    model: SignalModel
    raw_signal: str
    normalized_score: float = Field(..., ge=-1.0, le=1.0)
    confidence: float = Field(1.0, ge=0.0, le=1.0)
    signal_label: Optional[str] = None
    source: str


class EnsembleDecision(BaseModel):
    date: date
    close: float = Field(..., gt=0.0)
    action: TradeAction
    average_score: float = Field(..., ge=-1.0, le=1.0)
    target_exposure: Optional[float] = Field(None, ge=0.0, le=2.5)
    model_count: int = Field(..., ge=1, le=3)
    model_scores: Dict[str, float]
    sentiment_action: Optional[TradeAction] = None
    support_score: float = Field(0.0, ge=-1.0, le=1.0)
    technical_adjustment: float = Field(0.0, ge=-2.5, le=2.5)
    fundamental_adjustment: float = Field(0.0, ge=-2.5, le=2.5)


class EnsembleTrade(BaseModel):
    date: date
    action: TradeAction
    price: float = Field(..., gt=0.0)
    exposure_before: float
    exposure_after: float
    trade_value: float
    transaction_cost: float
    shares_after: float
    cash_after: float
    portfolio_value: float


class EnsembleEquityPoint(BaseModel):
    date: date
    close: float = Field(..., gt=0.0)
    shares: float
    cash: float
    exposure: float
    portfolio_value: float
    daily_return: float


class EnsembleBacktestMetrics(BaseModel):
    initial_capital: float
    final_value: float
    total_return_pct: float
    sharpe_ratio: float
    max_drawdown_pct: float
    win_rate_pct: float
    trade_count: int
    decision_count: int
    average_model_count: float
    coverage_by_model: Dict[str, int]


class EnsembleBacktestResponse(BaseModel):
    ticker: str
    market: str
    start_date: date
    end_date: date
    metrics: EnsembleBacktestMetrics
    decisions: List[EnsembleDecision]
    trades: List[EnsembleTrade]
    equity_curve: List[EnsembleEquityPoint]
    model_signals: List[EnsembleModelSignal]
    source_files: List[str]
    warnings: List[str]
    generated_at: datetime
